src/policy/immix/immixspace.rs

```diff
@@ -911,6 +910,17 @@ impl<VM: VMBinding> GCWork<VM> for SweepChunk<VM> {
                     // Block is live. Increment the allocated block count.
                     allocated_blocks += 1;
                 }
+
+                // Clear side forwarding bits.
+                // In the beginning of the next GC, no side forwarding bits shall be set.
```

Comment:
Is it possible to assert this? We probably don't need the following line, and we can add some assertions there: https://github.com/mmtk/mmtk-core/blob/170630fc84bc82070ca6e06ea4fed76894aa703f/src/policy/immix/immixspace.rs#L631

Reply:
It is possible to assert that there is no forwarding bits set in the beginning of a GC.  But that'll be extreme costly.  For the release build, it is unnecessary; for the debug build, such array/memory traversal operation is especially costly due to lack of optimisations including inlining.

The invocations of `clear_forwarding_bits` on line 622 and line 631 are unrelated to this.  It is called when one GC thread won the forwarding race, but later found that it cannot forward the object because the object is already marked (line 622) or the object is pinned or the defrag space exhausted (line 631).  It needs to revert it to the `FORWARDING_NOT_TRIGGERED_YET` state.  If the state were left in the `BEING_FORWARDED` state, other GC workers will block on this object forever, waiting for it to be forwarded, which will never happen; if we change it to `FORWARDED` state, other GC workers will attempt to load from its forwarding pointer which is never written.  So the state has to be `FORWARDING_NOT_TRIGGERED_YET`.  When another worker visits this object, it will won the forwarding race again, see it is already marked, and revert it to `FORWARDING_NOT_TRIGGERED_YET` again.

Reply:
I ran some tests locally with StickyImmix and OpenJDK with side forwarding bits, and observed that in nursery GC right after a defrag GC, a worker may observe the forwarding bits being set, but the forwarding pointer is 0.  The problem disappears if I remove the optimisation about defrag sources (i.e. force clearing forwarding bits for all blocks).  This is strange.  It looks like some forwarding bits are set in non-defrag-source blocks.

Reply:
It ended up that whether a block is a defrag source or not is determined by `DEFRAG_STATE_TABLE`, but `block.sweep()` overwrites `DEFRAG_STATE_TABLE` with the number of holes.  That caused my code to consider fewer blocks as defrag sources, and that's why the problem disappears if it clears forwarding bits regardless whether a block is a defrag source.  The fix is simply swapping the order of clearing forwarding bits and sweeping the block.

Reply:
There is another invocation of `clear_forwarding_bits::<VM>(object);` in `post_copy`.  I changed it to be executed unconditionally when I observed that the copy context may allocate the to-space copy into a free block with stale forwarding bits.  That was because we lazily clear the forwarding bits.  Now that we clear forwarding bits at the end of copy GCs, we are sure there must not be stale forwarding bits.  I will revert that commit and elide the `clear_forwarding_bits` if the forwarding bits are on the side.

Reply:
More precisely, it is re-applying https://github.com/mmtk/mmtk-core/commit/d341506b70a4465dbbc05e87d27d0b7bd23844bb which made that line conditional in the first place.